lib/js-api/src/instance.rs

```diff
@@ -0,0 +1,143 @@
+/// interacting with WebAssembly.
+///
+/// Spec: <https://webassembly.github.io/spec/core/exec/runtime.html#module-instances>
+#[derive(Clone)]
+pub struct Instance {
+    instance: WebAssembly::Instance,
+    module: Module,
+    /// The exports for an instance.
+    pub exports: Exports,
+}
+
+/// An error while instantiating a module.
+///
+/// This is not a common WebAssembly error, however
+/// we need to differentiate from a `LinkError` (an error
+/// that happens while linking, on instantiation), a
+/// Trap that occurs when calling the WebAssembly module
+/// start function, and an error when initializing the user's
+/// host environments.
+#[derive(Error, Debug)]
+pub enum InstantiationError {
+    /// A linking ocurred during instantiation.
+    #[cfg_attr(feature = "std", error("Link error: {0}"))]
+    Link(String),
+
+    /// A runtime error occured while invoking the start function
+    #[cfg_attr(feature = "std", error("Start error: {0}"))]
+    Start(String),
+    // /// Error occurred when initializing the host environment.
+    // #[error(transparent)]
+    // HostEnvInitialization(HostEnvInitError),
+}
+
+// impl From<wasmer_engine::InstantiationError> for InstantiationError {
+//     fn from(other: wasmer_engine::InstantiationError) -> Self {
+//         match other {
+//             wasmer_engine::InstantiationError::Link(e) => Self::Link(e),
+//             wasmer_engine::InstantiationError::Start(e) => Self::Start(e),
+//         }
+//     }
+// }
+
+// impl From<HostEnvInitError> for InstantiationError {
+//     fn from(other: HostEnvInitError) -> Self {
+//         Self::HostEnvInitialization(other)
+//     }
+// }
+
+impl Instance {
+    /// Creates a new `Instance` from a WebAssembly [`Module`] and a
+    /// set of imports resolved by the [`Resolver`].
+    ///
+    /// The resolver can be anything that implements the [`Resolver`] trait,
+    /// so you can plug custom resolution for the imports, if you wish not
+    /// to use [`ImportObject`].
+    ///
+    /// The [`ImportObject`] is the easiest way to provide imports to the instance.
+    ///
+    /// [`ImportObject`]: crate::ImportObject
+    ///
+    /// ```
+    /// # use wasmer_js::{imports, Store, Module, Global, Value, Instance};
+    /// # fn main() -> anyhow::Result<()> {
+    /// let store = Store::default();
+    /// let module = Module::new(&store, "(module)")?;
+    /// let imports = imports!{
+    ///   "host" => {
+    ///     "var" => Global::new(&store, Value::I32(2))
+    ///   }
+    /// };
+    /// let instance = Instance::new(&module, &imports)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// ## Errors
+    ///
+    /// The function can return [`InstantiationError`]s.
+    ///
+    /// Those are, as defined by the spec:
+    ///  * Link errors that happen when plugging the imports into the instance
+    ///  * Runtime errors that happen when running the module `start` function.
```

Comment:
Nah, those aren't returned if I'm following the code correctly. Can you confirm?

Reply:
Let me fix that

Reply:
Fixed!